### PR TITLE
replace product base_url with versions object

### PR DESCRIPTION
**Pull request by danawillow**
First step towards #155. Adds the ability to set multiple versions, but doesn't do anything with that yet- for now, it just uses the default version for everything, so there should be no changes to downstream repos.

-----------------------------------------------------------------
# [all]

### CODE REVIEW

**Comment by danawillow** [comment]
Oh huh, the magician also failed because of the same bundler/compiler thing. Not really sure what the issue is- past magician runs have been successful so it doesn't seem to be a problem beyond this PR.

**Comment by rambleraptor** [comment]
Maybe try rebasing off of #159? That PR fixes all of the Rubocop + Rspec tests. It looks like Travis is failing because Rubocop/Rspec both fail.

**Comment by danawillow** [comment]
Eh, I'll merge that in once it gets submitted, that way I don't have to keep rebasing if there are changes. My concern was about the `build/run/presubmit` one.

**Comment by rosbo** [comment]
You can remove the git hook in your repo to avoid having the ChangeId added to the description. ChangeId was necessary for GoB but GoB is dead!

Run: `rm .git/hooks/commit-msg` at the root of your local checkout of MM

File: products/dns/api.yaml [hunk 1]
```diff
@@ -14,7 +14,10 @@
 --- !ruby/object:Api::Product
 name: Google Cloud DNS
 prefix: gdns
-base_url: https://www.googleapis.com/dns/v1/
+versions:
+  - !ruby/object:Api::Product::Version
+    name: v1
+    base_url: https://www.googleapis.com/dns/v1/
```
Comment (rosbo) [inline products/dns/api.yaml]: Shouldn't you put `default=true`? With the current logic, the compiler should have thrown an exception...

File: api/product.rb [hunk 2]
```diff
@@ -18,22 +18,59 @@
...
+    class Version < Api::Object
+      attr_reader :base_url
+      attr_reader :default
+      attr_reader :name
+
+      def validate
+        super
+        check_property :base_url, String
+        check_property :name, String
+        check_optional_property :default, [TrueClass, FalseClass]
```
Comment (rosbo) [inline api/product.rb]: Use `check_optional_property :default, :boolean`

File: api/product.rb [hunk 3]
```diff
@@ -18,22 +18,59 @@
...
+      end
+    end
+
+    def default_version
+      @versions.each do |v|
```
Comment (rosbo) [inline api/product.rb]: Consider adding `return @versions.last if @versions.length == 1` and adjusting the `check_versions` below. I am fine also with requiring default to be set to true even if we only have one version.

File: api/product.rb [hunk 4]
```diff
@@ -18,22 +18,59 @@
...
+    end
+
+    # Represents a version of the API for this product
+    class Version < Api::Object
+      attr_reader :base_url
+      attr_reader :default
+      attr_reader :name
+
+      def validate
+        super
```
Comment (rosbo) [inline api/product.rb]: add `@default ||= false` to set @default to false if the value is not specified. This way, you avoid propagating a `nil` value.

File: products/iam/api.yaml [hunk 5]
```diff
@@ -16,7 +16,10 @@
 --- !ruby/object:Api::Product
 name: Google Cloud IAM
 prefix: giam
-base_url: https://iam.googleapis.com/v1/
+versions:
+  - !ruby/object:Api::Product::Version
+    name: v1
+    base_url: https://iam.googleapis.com/v1/
```
Comment (rosbo) [inline products/iam/api.yaml]: `default=true`? and for all the other files too. Unless you implement the logic I suggested for `default_version`

File: api/product.rb (hunk 4 above)
Comment (danawillow) [reply api/product.rb]: Done.

File: api/product.rb (hunk 3 above)
Comment (danawillow) [reply api/product.rb]: Done.

File: products/dns/api.yaml (hunk 1 above)
Comment (danawillow) [reply products/dns/api.yaml]: No longer necessary.

File: api/product.rb (hunk 2 above)
Comment (danawillow) [reply api/product.rb]: Done

File: products/container/api.yaml [hunk 6]
```diff
@@ -14,7 +14,11 @@
 --- !ruby/object:Api::Product
 name: Google Container Engine
 prefix: gcontainer
-base_url: https://container.googleapis.com/v1/
+versions:
+  - !ruby/object:Api::Product::Version
+    name: v1
+    base_url: https://container.googleapis.com/v1/
+    default: true
```
Comment (rosbo) [inline products/container/api.yaml]: You could remove this line for consistency

**Pull request merged**